File: examples/localcowork/mcp-servers/_shared/ts/json-rpc.ts
/**
 * JSON-RPC 2.0 Transport Utilities — TypeScript
 *
 * Low-level JSON-RPC message handling for MCP server communication.
 * Used by mcp-base.ts; typically not imported directly by tool implementations.
 */

// ─── JSON-RPC Types ─────────────────────────────────────────────────────────

export interface JsonRpcMessage {
  jsonrpc: '2.0';
}

export interface JsonRpcRequest extends JsonRpcMessage {
  id: string | number;
  method: string;
  params?: Record<string, unknown>;
}

export interface JsonRpcNotification extends JsonRpcMessage {
  method: string;
  params?: Record<string, unknown>;
}

export interface JsonRpcSuccessResponse extends JsonRpcMessage {
  id: string | number;
  result: unknown;
}

export interface JsonRpcErrorResponse extends JsonRpcMessage {
  id: string | number;
  error: {
    code: number;
    message: string;
    data?: unknown;
  };
}

export type JsonRpcResponse = JsonRpcSuccessResponse | JsonRpcErrorResponse;

// ─── Helpers ────────────────────────────────────────────────────────────────

/** Create a success response */
export function successResponse(id: string | number, result: unknown): JsonRpcSuccessResponse {
  return { jsonrpc: '2.0', id, result };
}

/** Create an error response */
export function errorResponse(
  id: string | number,
  code: number,
  message: string,
  data?: unknown,
): JsonRpcErrorResponse {
  return { jsonrpc: '2.0', id, error: { code, message, data } };
}

/** Create a notification (no response expected) */
export function notification(method: string, params?: Record<string, unknown>): JsonRpcNotification {
  return { jsonrpc: '2.0', method, params };
}

/** Validate that a parsed message is a valid JSON-RPC request */
export function isValidRequest(msg: unknown): msg is JsonRpcRequest {
  if (typeof msg !== 'object' || msg === null) return false;
  const obj = msg as Record<string, unknown>;
  return (
    obj.jsonrpc === '2.0' &&
    typeof obj.method === 'string' &&
    (typeof obj.id === 'string' || typeof obj.id === 'number')
  );
}

/** Parse a raw string into a JsonRpcRequest, throwing on invalid input */
export function parseRequest(raw: string): JsonRpcRequest {
  const parsed = JSON.parse(raw);
  if (!isValidRequest(parsed)) {
    throw new Error('Invalid JSON-RPC request');
  }
  return parsed;
}

File: examples/localcowork/mcp-servers/_shared/py/json_rpc.py
"""
JSON-RPC 2.0 Transport Utilities — Python

Low-level JSON-RPC message handling for MCP server communication.
Used by mcp_base.py; typically not imported directly by tool implementations.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class JsonRpcRequest:
    """A JSON-RPC 2.0 request."""

    jsonrpc: str
    id: str | int
    method: str
    params: dict[str, Any] | None = None


@dataclass
class JsonRpcResponse:
    """A JSON-RPC 2.0 response."""

    jsonrpc: str = "2.0"
    id: str | int = 0
    result: Any = None
    error: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"jsonrpc": self.jsonrpc, "id": self.id}
        if self.error is not None:
            d["error"] = self.error
        else:
            d["result"] = self.result
        return d


def success_response(request_id: str | int, result: Any) -> dict[str, Any]:
    """Create a JSON-RPC success response."""
    return {"jsonrpc": "2.0", "id": request_id, "result": result}


def error_response(
    request_id: str | int,
    code: int,
    message: str,
    data: Any = None,
) -> dict[str, Any]:
    """Create a JSON-RPC error response."""
    error: dict[str, Any] = {"code": code, "message": message}
    if data is not None:
        error["data"] = data
    return {"jsonrpc": "2.0", "id": request_id, "error": error}


def notification(method: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
    """Create a JSON-RPC notification (no id, no response expected)."""
    msg: dict[str, Any] = {"jsonrpc": "2.0", "method": method}
    if params is not None:
        msg["params"] = params
    return msg


def is_valid_request(msg: dict[str, Any]) -> bool:
    """Validate that a parsed dict is a valid JSON-RPC 2.0 request."""
    return (
        isinstance(msg, dict)
        and msg.get("jsonrpc") == "2.0"
        and isinstance(msg.get("method"), str)
        and ("id" in msg and isinstance(msg["id"], (str, int)))
    )
